Initial minimal logins API connection to datastore

lmorchard opened the pull request:
Connected to #21

This doesn't actually fix #21, but it's kind of a start as it connects up the add-on to the real logins info data.

## Testing and Review Notes

* The add-on should now display real login entries
* Editing the login entries from the add-on should affect changes in the in-product login manager.
* Editing entries in the in-product login manager should affect changes in the add-on.

lmorchard commented on src/list/reducers.js:
```diff
@@ -37,7 +37,13 @@ export function cacheReducer(state = {items: [], currentItem: null}, action) {
       ...state,
-      items: [...state.items, makeItemSummary(action.item)],
+      items: [
+        // HACK: add dispatched via Logins API events may be a duplicate
```
Small hacky wrinkle here: Since I'm broadcasting datastore changes based on Logins API events without any source, the UI source that made the changes will get its own change dispatched back to it. This will look like a duplicate. So, I papered over that by making the "add" action idempotent here.

lmorchard commented:
Okay, I *think* I worked out the issues. At least, insofar as I'm going to take off the "WIP" title and go get some lunch.

lmorchard commented:
Oh, but one other thing I may need to investigate for this PR or in a follow-up: Nothing here calls `touch()` on the Logins API to register that a login has been used. Wanted to do a little more digging to make sure I don't make fetches for display end up counting as a use.

6a68 commented on src/background/datastore.js:
```diff
@@ -133,13 +134,16 @@ class DataStore {
       timeCreated: Date.now(),
       timePasswordChanged: Date.now(),
     };
+    await browser.experiments.logins.add(added);
```
This and other logins-modifying API methods do throw if something goes wrong, so I'd wrap them in try/catch blocks and just log the errors for now. I think we can handle the errors in the relevant issues (looks like #18, #30, #34 might all be relevant).

lmorchard replied:
Hmm, I can just silently log them, though datastore.js *does* throw errors for other validation problems and such. Kind of figured it was useful to let these throw too, though it doesn't look like we do anything useful with them yet. (e.g. report an error over a message port, etc)

6a68 replied:
> datastore.js does throw errors for other validation problems and such

Ah, ok. Probably fine as is, then.

6a68 commented on src/background/datastore.js:
```diff
@@ -133,13 +134,16 @@ class DataStore {
       timePasswordChanged: Date.now(),
     };
+    await browser.experiments.logins.add(added);
     this._all[added.guid] = added;
```
I might be getting a little ahead of this transitional PR, but it seems cleaner to remove this line, so that `add()` just lobs the new login over to Firefox, and the datastore relies solely on the `onAdded` listener to update the list of logins.

lmorchard replied:
Yeah, that kind of feels like in the spirit of uni-directional dataflow ala redux, even if we're not using it here :)

linuxwolf replied:
Flux is a way of life (-:

6a68 commented on src/background/datastore.js:
```diff
@@ -159,11 +163,11 @@ class DataStore {
       timePasswordChanged: Date.now(),
     };
+    await browser.experiments.logins.update(updated);
     this._all[updated.guid] = updated;
```
Same comments as `add()` on these two lines (try/catch and maybe not directly updating the internal datastore).

6a68 commented on src/background/datastore.js:
```diff
@@ -174,20 +178,64 @@ class DataStore {
     if (item) {
-      // TODO: call API
+      await browser.experiments.logins.remove(item.id);
       delete this._all[item.id];
```
Same comments as above for this spot, too.

6a68 commented on test/unit/background/datastore-test.js:
```diff
@@ -0,0 +1,266 @@
+
+const cmpAlphaBy = name => (a, b) => a[name].localeCompare(b[name]);
+
+const SAMPLE_INFOS = {
```
For completeness' sake, probably worth including the rest of the [`nsILoginMetaInfo`](https://searchfox.org/mozilla-central/source/toolkit/components/passwordmgr/nsILoginMetaInfo.idl) fields in these mocks: `timeCreated`, `timeLastUsed`, `timePasswordChanged`, `timesUsed`.

linuxwolf commented:
> Nothing here calls `touch()` on the Logins API to register that a login has been used

That was on the list of things to get to from before; I'll make sure it's accounted for in our issues.

lmorchard commented on src/background/message-ports.js:
```diff
@@ -44,19 +44,16 @@ export default function initializeMessagePorts() {
     case "add_item":
       return openDataStore().then(async (ds) => {
         const item = await ds.add(message.item);
-        broadcast({type: "added_item", item}, sender);
```
Realized that since I've got the datastore now broadcasting these on the Logins API event handlers, all these messages got doubled up.

lmorchard commented on test/unit/background/message-ports-test.js:
```diff
@@ -80,7 +83,6 @@ describe("background > message ports", () => {
     itemId = result.item.id;
 
     expect(result.item).to.deep.include(item);
-    expect(selfListener).to.have.callCount(0);
```
Since the datastore broadcasts messages to relay events from the Logins API, there's no sender or self available at that time to refrain from sending the event back to "self"

lmorchard commented on test/unit/mocks/browser.js:
```diff
@@ -190,4 +194,22 @@ window.browser = {
+    logins: {
+      async getAll() { return []; },
+      async add(login) {
+        browser.experiments.logins.onAdded.getListener()({ login });
```
I was hoping I could be less verbose here by just using `this`, but that apparently ends up being the wrong reference

linuxwolf replied:
🤔

lmorchard commented:
Addressed comments except for wrapping things in try/catch to log errors. We might want to let those throw, so we can catch those higher up along with datastore validation errors.

6a68 commented:
Nice work :+1: Took this branch for a test drive, and all the logins add / update / delete happy paths are seamlessly mirrored across the logins management UI in Firefox, and the Lockbox management UI.

I dunno if @linuxwolf wants to take a pass before merging, but LGTM
